chefkoch/recipe.py

```diff
@@ -0,0 +1,771 @@
+                return str("The recipe contains a circle reachable from " +
+                           node.name)
+        return ""
+
+    def recursiveDFS(self, node, nodesOnTheWay):
+        """
+        Recursive Depth First Search finding circles.
+        
+        :param node:            The node, the DFS starts in.
+        :type node:             Node
+        :param nodesOnTheWay:   Previously visited nodes. If a node in there can
+                                be visited by going deeper into the graph, \
+                                there is a circle.
+        :type nodesOnTheWay:    Node[]
+        :returns:               `True` if there is a circle. False elsewise.
+        """
+        namesOnTheWay = ""
+        for nodeOTW in nodesOnTheWay:
+            namesOnTheWay = namesOnTheWay + " " + nodeOTW.name
+        logger.debug("Executing rDFS for " + node.name + " and " +
+                     namesOnTheWay)
+        if node in nodesOnTheWay:
+            logger.warning("The recipe contains a circle along " +
+                        namesOnTheWay + node.name + " and can therefore" +
+                        " not be executed.")
+            return True
+        nodesOnTheWay.append(node)
+        # this only gets longer in deeper recursion levels
+        # no need to bring nodes from deeper levels up.
+        for key in node.outputs:
+            output = node.outputs[key]
+            # output might be used by several other nodes
+            for nextNode in self.nodes:
+                # to search for the values in a dict instead of their keys
+                # we need to invert it
+                invertedInputDict = dict(map(reversed, nextNode.inputs.items()))
+                if output in invertedInputDict:
+                    logger.debug("Taking the edge from " + node.name + " to " +
+                                 nextNode.name)
+                    if self.recursiveDFS(nextNode, nodesOnTheWay):
+                        return True # a circle was found
+                    # else continue with next node
+        # after all outgoing edges where tested, remove current node
+        nodesOnTheWay.remove(node)
+        # if there is no circle from none of the outputs
+        return False
+
+
+class Node:
+    """
+    A node encapsules a simulation step within a recipe. The step can
+    be realised by a python file, a sub-recipe or a built-in function.
+    Each node also has a name, a dict of inputs and a dict of outputs.
+    To the inputdict, the key is the same input name that the step takes,
+    the value is where the input comes from. To the outputdict, the key
+    is the name the step uses, the value is the name under which the
+    output is available to other nodes in the recipe (the same name used
+    as value in another inputdict).
+    """
+    def __init__(self, name, inputdict, outputdict, stepsource):
+        """
+        Initializes a node of the recipe. A node represents a simulation
```

Comment:
As a general rule (also applying to the commit messages): Please formulate this first line as a command that characterizes the function/commit.

Example:
```
Initialize a node of the Recipe.

A node of the recipe represents a simulation step.
```

Further, decide which parameter representation you want to use. If you stick to `sphinx-numpydoc`, apply the following scheme:
```
'''
Parameters
------------
    param_name : str
        A description of this parameter.
    another_thingy : int
        Some random parameter. Chosen by fair dice roll.
    some_kw_argument : str, optional
        Some keyworded argument.

        Defaults to 'none'.

Returns
--------
    int
        The output is a number.
```

Then, this will be layouted nicely by sphinx-numpydoc.

This remarks generally applies to all commits and docstrings throughout the package. Please check